src/algebra/regular.lean

```diff
@@ -0,0 +1,241 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+
+/-!
+# Regular elements
+
+We introduce left-regular, right-regular and regular elements.
+
+By definition, a regular element in a commutative ring is a non-zero divisor.
+Lemma `is_regular_of_cancel_monoid_with_zero` implies that every non-zero element of an integral
+domain is regular.
+
+The lemmas in Section `mul_zero_class` show that the `0` element is (left/right-)regular if and
+only if the `mul_zero_class` is trivial.  This is useful when figuring out stopping conditions for
+regular sequences: if `0` is ever an element of a regular sequence, then we can extend the sequence
+by adding one further `0`.
+
+The final goal is to develop part of the API to prove, eventually, results about non-zero-divisors.
+-/
+variables {R : Type*} {a b : R}
+
+section has_mul
+
+variable [has_mul R]
+
+/-- A left-regular element is an element `c` such that multiplication on the left by `c`
+is injective on the left. -/
+def is_left_regular (c : R) := function.injective ((*) c)
+
+/-- A right-regular element is an element `c` such that multiplication on the right by `c`
+is injective on the right. -/
+def is_right_regular (c : R) := function.injective (* c)
+
+/-- A regular element is an element `c` such that multiplication by `c` both on the left and
+on the right is injective. -/
+structure is_regular (c : R) : Prop :=
+(left : is_left_regular c)
+(right : is_right_regular c)
+
+end has_mul
+
+section semigroup
+
+variable [semigroup R]
+
+/-- In a semigroup, then the product of left-regular elements is left-regular. -/
+lemma is_left_regular.mul (lra : is_left_regular a) (lrb : is_left_regular b) :
+  is_left_regular (a * b) :=
+show function.injective ((*) (a * b)), from (comp_mul_left a b) ▸ lra.comp lrb
+
+/-- In a semigroup, then the product of right-regular elements is right-regular. -/
+lemma is_right_regular.mul (rra : is_right_regular a) (rrb : is_right_regular b) :
+  is_right_regular (a * b) :=
+show function.injective (* (a * b)), from (comp_mul_right b a) ▸ rrb.comp rra
+
+/--  If an element `b` becomes left-regular after multiplying it on the left by a left-regular
+element, then `b` is left-regular. -/
+lemma is_left_regular_of_mul_left (ab : is_left_regular (a * b)) :
+  is_left_regular b :=
+function.injective.of_comp (by rwa comp_mul_left a b)
```

Comment:
```suggestion
lemma is_left_regular.of_mul (ab : is_left_regular (a * b)) :
  is_left_regular b :=
function.injective.of_comp (by rwa comp_mul_left a b)
```
we may as well copy the naming scheme of `injective` here

Reply:
Eric, your previous comments came between a change and the next one!

For this one, how does it play with the `right` version?  the `of_mul` stays the same, and does not specify which side the product is?

Reply:
Which `right` version? I only see a lemma I'd call `is_right_regular.of_mul`

Reply:
Yes, there are these two lemmas exactly.  My question is about the fact that the two `of_mul` refer once to `mul_left` and once to `mul_right`, even though the name makes no reference to this.  I am not too worried by this, but simply wanted to confirm!

Reply:
We have the same duality with `surjective.of_comp` and `injective.of_comp`, with the composition being from different sides - so I think it's probably ok. If it makes you uneasy, then maybe we should rename those lemmas too

Reply:
Eric, thanks for the clarification!  I updated the names, aligning them with the convention for injective.  I do not mind too much about not recording the asymmetry: I am very likely going to use these lemmas mostly in the commutative case anyway.